### PR TITLE
net: lib: nrf_cloud: implement uninit functionality

### PR DESCRIPTION
**Pull request by jayteemo**
Allow the nrf cloud module to be uninitialized (if no FOTA job is active).
Cleaned up some state handling also.
Ref: CIA-346

Signed-off-by: Justin Morton <justin.morton@nordicsemi.no>

### CODE REVIEW

File: include/net/nrf_cloud.h [hunk 1]
```diff
@@ -229,8 +229,8 @@ struct nrf_cloud_init_param {
 /**
  * @brief Initialize the module.
  *
- * @warning This API must be called exactly once, and it must return
- * successfully.
+ * @warning  This API must be called prior to using nRF Cloud
```
Comment (plskeggs) [inline include/net/nrf_cloud.h:232]: Is there now an extra space between @warning and This?

File: include/net/nrf_cloud.h [hunk 2]
```diff
@@ -239,6 +239,20 @@ struct nrf_cloud_init_param {
...
 int nrf_cloud_init(const struct nrf_cloud_init_param *param);
 
+/**
+ * @brief Uninitialize nRF Cloud; disconnects cloud connection
+ *  and cleans up allocated memory.
+ *
+ * @note If nRF Cloud FOTA is enabled and a FOTA job is active
+ *  uninit will not be performed.
+ *
+ *
```
Comment (plskeggs) [inline include/net/nrf_cloud.h:249]: Remove extra empty line?

**Review by plskeggs — APPROVED** [review]
Looks good

File: include/net/nrf_cloud.h (hunk 1 above)
Comment (jayteemo) [reply include/net/nrf_cloud.h:232]: there... was.

File: include/net/nrf_cloud.h [hunk 3]
```diff
@@ -239,6 +239,19 @@ struct nrf_cloud_init_param {
...
 
+/**
+ * @brief Uninitialize nRF Cloud; disconnects cloud connection
+ *  and cleans up allocated memory.
+ *
+ * @note If nRF Cloud FOTA is enabled and a FOTA job is active
+ *  uninit will not be performed.
+ *
+ * @retval 0      If successful.
+ *         -EBUSY If a FOTA job is in progress.
```
Comment (rlubos) [inline include/net/nrf_cloud.h:250]: AFAIK each possible return value should have a separate `@retval` tag.

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_fota.h [hunk 4]
```diff
@@ -119,6 +119,16 @@ typedef void (*nrf_cloud_fota_ble_callback_t)
  */
 int nrf_cloud_fota_init(nrf_cloud_fota_callback_t cb);
 
+/**
+ * @brief Uninitialize nRF Cloud FOTA; cleans up allocated memory. If a FOTA
+ *  job is in progress, uninit is not be performed and an error status is
+ *  returned.
+ *
+ * @retval 0      If successful.
+ *         -EBUSY If a FOTA job is in progress.
```
Comment (rlubos) [inline subsys/net/lib/nrf_cloud/include/nrf_cloud_fota.h:128]: AFAIK each possible return value should have a separate @retval tag.

**Review by rlubos — APPROVED** [review]
Just some doc nits, looks good otherwise.

File: include/net/nrf_cloud.h (hunk 3 above)
Comment (jayteemo) [reply include/net/nrf_cloud.h:250]: i did not know this, thanks!

**Pull request merged**